### PR TITLE
Homework 02

### PR DESCRIPTION
**Pull request by Fle9ma**
# Assignment
- [x] Create a configuration that will provide this class as a dependency. Value for each field should be provided through `.properties/.yml` file.
- [x] Inject `DataSource` dependency in the implementation of the in-memory repository and just print all the values in the `init {}` block.
- [x] For in-file repository implementation you will have to provide `Resource` dependency - specifically file resource that will point to any path in your OS' file system. For the file name, use `dbName` property of the `DataSource` class.
- [x] Create a main method that will instantiate `ApplicationContext`, fetch a service you’ve implemented and use it to manually test you application (manually call few methods you have defined in the service)
- [x] Write a couple of unit and integrations tests that will verify application behavior. Try to cover as much use-cases and corner cases as possible. Make sure you override defined `.properties` with different values so that `DataSource` injected in tests will have different values

# Optional
- [x] Play around with properties and bean creation a bit more, add property that will be a "switch" - when it’s "turned-on", application will use in-memory repository and when it’s "turned off", application will use in-file repository. You can then manually set value of the "switch" to provide a specific repository implementation
- [x] Add ktlint to the project, a library that formats your code. You can use Gradle plugin to easily integrate it with your project. Before you commit your code, try running ktlint task `ktlintFormat` to tidy up your code.

### CODE REVIEW

File: kotlin_intro/src/test/kotlin/com/infinum/homework/service/CarCheckUpServiceUnitTests.kt [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        assertThat(service.isCheckUpNecessary(cars[1].vin)).isTrue
+        verify(exactly = 1) { repository.findAll() }
+    }
+
+    @Test
+    fun isCheckUpNecessaryCarNotFound() {
+        every { repository.findAll().values } returns listOf(
+            CarCheckUp(1, LocalDateTime.of(2020, Month.MAY, 25, 13, 30), cars[0])
+        )
+        assertThat(service.isCheckUpNecessary("JH4DA9470MS008981")).isFalse
```
Comment (ivanmiklec) [inline kotlin_intro/src/test/kotlin/com/infinum/homework/service/CarCheckUpServiceUnitTests.kt:73]: Ovaj test faila jer ovdje moraš eksplicitno assertat da case baca exception:
```suggestion
 assertThatThrownBy { service.isCheckUpNecessary("JH4DA9470MS008981") }
            .isInstanceOf(NoVinException::class.java).hasMessage("Car with VIN JH4DA9470MS008981 does not exist")
```

File: kotlin_intro/src/main/kotlin/com/infinum/homework/CarCheckUpApplication.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.core.io.FileSystemResource
+
+@Configuration
+@ComponentScan
+@PropertySource("classpath:application.properties")
+class ApplicationConfiguration {
+
+    @Bean
+    fun fileResource(@Value(value = "src/main/resources/\${data-source.dbName}") path: String) =
+        FileSystemResource(path)
```
Comment (ivanmiklec) [inline kotlin_intro/src/main/kotlin/com/infinum/homework/CarCheckUpApplication.kt:17]: Ovdje čak ne moraš kreirat svoj bean, ako ga injectas u servis, automatski će inicijalizirat resurs, ovo je dobar članak na temu toga: https://www.logicbig.com/tutorials/spring-framework/spring-core/injecting-resources.html

File: kotlin_intro/src/main/kotlin/com/infinum/homework/repository/InFileCarCheckUpRepository.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.core.io.Resource
+import org.springframework.stereotype.Repository
+import java.io.FileOutputStream
+import java.time.LocalDateTime
+
+@Repository
+@Qualifier("inFileCarCheckUpRepository")
+@ConditionalOnProperty(name = ["repository.switch"], havingValue = "turned-off")
+class InFileCarCheckUpRepository(private val carCheckUpsFileResource: Resource) :
```
Comment (ivanmiklec) [inline kotlin_intro/src/main/kotlin/com/infinum/homework/repository/InFileCarCheckUpRepository.kt:16]: Nastavno na komentar vezano uz resource, ovo ti onda samo postane:
```suggestion
class InFileCarCheckUpRepository(@Value("\${data-source.dbName}") private val carCheckUpsFileResource: Resource) :
```

Također onda kao property vrijednost za `data-source.dbName` u application.properties možeš iskoristit keyword `classpath` u springu koji će ti odmah pokazat na `resource` direktorij, pa to onda bude ovakva vrijednost na kraju: `classpath:car_checkups.txt`

File: kotlin_intro/src/main/kotlin/com/infinum/homework/service/CarCheckUpService.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        return if (lastCheckUp == null) {
+            true
+        } else {
+            val lastCheckUpDate: LocalDateTime = lastCheckUp.performedAt
+            val now: LocalDateTime = LocalDateTime.now()
+            val period = lastCheckUpDate.until(now, ChronoUnit.YEARS)
+            period >= 1
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Fle9ma) [reply kotlin_intro/src/main/kotlin/com/infinum/homework/service/CarCheckUpService.kt:50]: Ty 👍 Super metoda, dosta olaksava stvar

File: kotlin_intro/src/test/kotlin/com/infinum/homework/service/CarCheckUpServiceIntegrationTests.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    fun isCheckUpNecessaryForCar() {
+        assertThat(service.isCheckUpNecessary(cars[0].vin)).isFalse
+        assertThat(service.isCheckUpNecessary(cars[1].vin)).isTrue
+        assertThat(service.isCheckUpNecessary(cars[2].vin)).isTrue
+        assertThat(service.isCheckUpNecessary(cars[3].vin)).isFalse
+    }
+
+    @Test
+    fun isCheckUpNecessaryForCarWithInvalidVin() {
```
[the remark this answers is not in the document]
Comment (Fle9ma) [reply kotlin_intro/src/test/kotlin/com/infinum/homework/service/CarCheckUpServiceIntegrationTests.kt:59]: Ty za link, napravim renaming svih metoda 🙌

File: kotlin_intro/src/test/kotlin/com/infinum/homework/service/CarCheckUpServiceUnitTests.kt (hunk 1 above)
Comment (Fle9ma) [reply kotlin_intro/src/test/kotlin/com/infinum/homework/service/CarCheckUpServiceUnitTests.kt:73]: Ty 🙌

File: kotlin_intro/src/main/kotlin/com/infinum/homework/repository/InFileCarCheckUpRepository.kt (hunk 3 above)
Comment (Fle9ma) [reply kotlin_intro/src/main/kotlin/com/infinum/homework/repository/InFileCarCheckUpRepository.kt:16]: Ovo je super, samo u ovakvoj implementaciji sam morao rucno stvoriti car_checkups.txt buduci da ako ga nema odmah baca Exception kroz constructor. Pa mi onda ovaj ispod if za provjeru da li taj file postoji ili ne nema nekog smisla.
Mozda sam nesto propustio pa slobodno bacis neki trik kako da rijesim trigeranje tog Exceptiona ako taj file vec nije kreiran.

Comment (ivanmiklec) [reply kotlin_intro/src/main/kotlin/com/infinum/homework/repository/InFileCarCheckUpRepository.kt]: U pravu si tu, tu sam ja propustio da file već mora bit kreiran kad se ide tom metodom, slobodno zadrži prethodnu implementaciju 👌
